lib/analysis/modules/ies_enkf.c

```diff
@@ -0,0 +1,1059 @@
+  data->ies_logfile          = NULL ;
+  data->ies_debug            = false;
+  data->max_gauss_newton_it  = 0;
+  data->state_size           = 0;
+  data->ens_mask0            = NULL;
+  data->ens_mask             = NULL;
+  data->obs_mask0            = NULL;
+  data->obs_mask             = NULL;
+  data->W                    = NULL;
+  data->A0                   = NULL;
+  data->E                    = NULL;
+  data->converged            = false;
+  data->config               = ies_enkf_config_alloc();
+  data->AAprojection         = true;
+  return data;
+}
+
+void ies_enkf_data_free( void * arg ) {
+  ies_enkf_data_type * data = ies_enkf_data_safe_cast( arg );
+  ies_enkf_config_free( data->config );
+  free( data );
+}
+
+void teccost(matrix_type * W,
+             matrix_type * D,
+             const char * fname,
+             int ens_size,
+             int izone)
+{
+   FILE *fp2;
+   float costJ[100];
+
+   if (izone == 1){
+      fp2 = fopen(fname, "w");
+      fprintf(fp2, "TITLE = \"%s\"\n",fname);
+      fprintf(fp2, "VARIABLES = \"i\" \"Average J\" ");
+      for (int i = 0; i < ens_size; i++)
+         fprintf(fp2, "\"%d\" ",i);
+      fprintf(fp2,"\n");
+   } else {
+      fp2 = fopen(fname, "a");
+   }
+
+   float costf=0.0;
+   for (int i = 0; i < ens_size; i++){
+      costJ[i]=matrix_column_column_dot_product(W,i,W,i)+matrix_column_column_dot_product(D,i,D,i);
+      costf += costJ[i];
+   }
+   costf= costf/ens_size;
+
+   fprintf(fp2,"%2d %10.3f ", izone-1, costf) ;
+   for (int i = 0; i < ens_size; i++)
+      fprintf(fp2,"%10.3f ", costJ[i] ) ;
+   fprintf(fp2,"\n") ;
+   fclose(fp2);
+}
+
+void teclog(matrix_type * W,
+            matrix_type * D,
+            matrix_type * DW,
+            const char * fname,
+            int ens_size,
+            int itnr,
+            double rcond,
+            int nrsing,
+            int nrobs)
+{
+
+   double diff1W = 0.0;
+   double diff2W = 0.0;
+   double diffW;
+   double costfp,costfd,costf;
+   FILE *fp;
+
+   if (itnr == 1){
+      fp = fopen(fname, "w");
+      fprintf(fp, "TITLE = \"%s\"\n",fname);
+      fprintf(fp, "VARIABLES = \"it\" \"Diff1W\" \"Diff2W\" \"J-prior\" \"J-data\" \"J\" \"rcond\" \"Sing\" \"nrobs\"\n");
+   } else {
+      fp = fopen(fname, "a");
+   }
+
+   for (int i = 0; i < ens_size; i++){
+       diffW=matrix_column_column_dot_product(DW,i,DW,i)/ens_size;
+       diff2W+=diffW;
+       if (diffW > diff1W) diff1W=diffW ;
+   }
+   diff2W=diff2W/ens_size;
+
+   costfp=0.0;
+   costfd=0.0;
+   for (int i = 0; i < ens_size; i++){
+       costfp += matrix_column_column_dot_product(W,i,W,i);
+       costfd += matrix_column_column_dot_product(D,i,D,i);
+   }
+   costfp=costfp/ens_size;
+   costfd=costfd/ens_size;
+   costf= costfp + costfd;
+   fprintf(fp," %d %12.5e %12.5e %12.5e %12.5e %12.5e %12.5e %d %d\n", itnr-1, diff1W, diff2W, costfp, costfd, costf, rcond, nrsing , nrobs ) ;
+   fclose(fp);
+}
+
+void tecfld(matrix_type * W,
```

Comment:
Some general comments:

1. This function (probably along with many of the others here ...) are clearly only meant to be used *inside* this file; then they should be marked as `static`.
2. The `matrix` argument should not be modified inside this function - specify it as `const matrix_type *W`.